docs/upgrade.md

```diff
@@ -0,0 +1,51 @@
+# Upgrade
+
+Currently the upgrade can be performed just between follow up downstream builds
+e.g. 171 -> 172.
+There is opened [BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1767400) which
+tracking the issue which prevents us to upgrade to the latest available build.
+
+There is prepared config option `upgrade_to_latest` in the DEPLOYMENT section
+which can be set or will be the default value when the BZ gets resolved.
+
+## Deployment
+
+To be able upgrade your cluster you cannot deploy the latest downstream
+build, because then you won't have any next build available for upgrade.
+
+For this purpose we have this config file: [upgrade.yaml](/conf/ocsci/upgrade.yaml)
+which you can pass via `--ocsci-conf` when you are running the deployment. This will
+automatically take not latest build but one before.
+
+## Upgrade execution
+
+For upgrade we have few pytest marks defined [here](/ocs_ci/framework/pytest_customization/marks.py):
+
+* **before_upgrade** - mark tests which are suppose to run before upgrade
```

Comment:
I think we should remove both `before_upgrade` and `after_upgrade` until there is full clarity on what the usecase is,  Right now most of us will need -m upgrade to upgrade from existing cluster and run our tests.  The additional marks will create confusions. We can add them later when really needed.

Reply:
In RHV we had much more complex ordering for upgrade tests. This is is basically prepared infrastructure which allow ppl to write upgrade related tests which will be executed in one jenkins upgrade job. As before upgrade test you can create some pod and after upgrade you can check if it survived the upgrade and so on.

I guess that @RazTamir will want ppl write such tests later on.

Reply:
And later its better to add some cli to fit our usecase ex:  `--ocs-version some_GA_version -m upgrade`(to upgrade to latest nightly)

Reply:
@petr-balogh I am not sure I got all the point, with cli option, its easier to do whatever upgrade one wants to run and whatever sequence they prefer. We can definitely discuss on this one before just following what you did in RHV.

Reply:
Both of you correct.
As we still don't have an actual (and official) way to test upgrades, it might be too early to introduce those new marks.  But it will come fast so it may be worth doing it in this PR.

One of the most important items for upgrade testing is to test that the existing cluster was not impacted by the upgrade process.
So we can have a small portion of tests dedicated to upgrading so we will be able to tell if something bad happened after the upgrade

For those tests we will need to ensure:
- Every upgrade test is combined from 2 tests - before upgrade and after upgrade
   * Before upgrade - create some entities with specific names (maybe IDs?)
   * After upgrade - use those entities by their names (maybe IDs?) and verify all good post upgrade
- Environment check is disabled
- NOT to teardown after the before upgrade test finished
- Teardown the after upgrade test


Every suggestion is welcomed of course as this approach might fit good for RHV like products but in K8s style, we might need to test more aspects

Reply:
Here is my input using cli and which will achieve the above use case without any additional marks:

Approach1: Using 2 pipelines:
pipeline1:
`run-ci -m deployment,some_upgrade_tests -ocs-version 4.x_GA`

pipeline2:
`run-ci -m upgrade,verify_upgrade_tests --ocs-version latest_nightly`

Approach 2(Just one pipeline):
 The test cases knows all about upgrades and does the job - better approach
Add upgrade tests in folder `/upgrade`
`run-ci -m upgrade, --ocs-version 4.2 --ocs-upgrade-version --nightly tests/upgrade`

Reply:
Let's discuss this on stuff/automation mtg, we currently have already the way how we specify the build which is deployed. It's via config [ocs_registry_image](https://github.com/red-hat-storage/ocs-ci/blob/master/ocs_ci/framework/conf/default_config.yaml#L49) . 

Currently upgrades cannot be performed to the latest nightly from some random previous version but it hast to be just one follow up build as described in the [BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1767400)

So now the upgrade is done just to one build above the current. If it will be supported now we are able to do just with changing one parameter [here](https://github.com/red-hat-storage/ocs-ci/pull/960/files#diff-3c73c8d9493c132ee1cabd09de37c01aR8) 

Anyway we still have just one version 4.2 and upgrading between internal builds and if user want just upgrade they can easily run just `run-ci -m upgrade` and that's it and it will upgrade the cluster.

I think that we are going to use branches for ocs-ci when we will have 4.3, this we also have to discuss which also changing the approach and you won't be able to perform upgrade just in one execution if you need to checkout to different branch between two stages 1) before upgrade,  2) upgrade and after upgrade tests.

Will be better if we will discuss all those things in the call.

Reply:
I am resolving this for now because still unclear at this stage with only one version.